src/v/kafka/server/handlers/produce.cc

```diff
@@ -348,9 +356,18 @@ produce_topic(produce_ctx& octx, produce_request::topic& topic) {
             continue;
         }
 
+        if (materialized) {
+            partitions_dispatched.push_back(ss::now());
+            partitions_produced.push_back(
+              ss::make_ready_future<produce_response::partition>(
+                produce_response::partition{
+                  .partition_index = part.partition_index,
+                  .error_code = error_code::invalid_topic_exception}));
+            continue;
+        }
```

Comment:
what happens if this check isn't here? presumably it would result in an error (e.g. because a partition lookup failed) and we could use that error rather than having a special case.

Reply:
I figured the error would be `partition_not_exists` which might give some confusion to the user. If the partition exists, but is materialized, giving an explicit error about why would be clearer. Thoughts?

Reply:
my feedback is less about the precise error message, and more about the code behaving sanely without the check. let it generate whatever error it normally does, and then catch and rewrite it.

Reply:
I think i understand the issue. The commit message is misleading: `prevent publishing onto materialized topics` as if that must be prevented otherwise the system would crash. Its actually unecessary, maybe i should just remove the commit altogether?

Reply:
the fast path should be producing to normal topics. so checking every produce to every partition as being destined to a materialized topic seems like overkill. instead, if we let the produce proceed and fail as a result of an existing condition that fails (e.g. partition does not exist) then this seems like a good trade off.

Reply:
Ok great, the commit has been removed